### ISSUE TITLE
Sile Build fails on FreeBSD 11

### ISSUE THREAD

**Issue by udsd007** [sile-typesetter/sile#425]
I have both Lua 5.1 and 5.2 installed, and all the Sile prerequisites installed for lua 5.1. I can't delete Lua 5.2, as it is required for vim; nor can I delete lua 5.1, as the prerequisites are not available for lua 5.2. 

I would very much like to get this installed and working. The attached file is the output of ./configure with no parameters: 

[configure.typescript.txt](https://github.com/simoncozens/sile/files/672498/configure.typescript.txt)

**Comment by alerque** [comment]
I'm not sure what it is off the top of my head, but there should be some variable to set or argument to pass `configure` to make it find a specific version of Lua. In your case your trouble starts here:

> checking for a Lua interpreter with version >= 5.1... lua52

Once it tries to use the latest version of Lua found on the system you run into trouble with the deps. You should be able to override this behavior by forcing it to find Lua 5.1.

That being said I'm wondering what's up with FreeBSD. Does in not have Lua 5.3? Why aren't other dependencies available across all Lua versions? All of SILE's dependencies work find on all versions of Lua supported (5.1, 5.2, 5.3) so this seems like a pickle you shouldn't be in if it weren't for some distro packaging problem. Maybe we can solve this at that level.

**Comment by simoncozens** [comment]
I'm looking into it. The problem is that the lua executable isn't linked with the Lua dynamic library, and I'm having trouble getting the linker to pick up the right library anyway.

**Comment by simoncozens** [comment]
Here's how I built it:

* pkg install git autoconf automake libtool pkgconf harfbuzz
* pkg install lua52 lua52-luarocks lua52-luafilesystem lua52-lpeg lua52-lzlib
* luarocks-5.2 install luaexpat
* ./configure CFLAGS="-I/usr/local/include" LDFLAGS="-L/usr/local/lib" LUA_LIB="-llua-5.2"
* make

**Comment by alerque** [comment]
@udsd007 Were you ever able to install and run SILE on FreeBSD? It looks to me like Simon's last comment has an answer to this question and it should be resolved so I'm going to close the issue for now, but please do report back either way whether this worked or now to confirm. And if there are any ongoing issues that were not resolved of course we'd like to hear about that!

**Comment by bsdzilla** [comment]
My two cents about this is, that with Simon's 26 Dec 2016 procedure, I was able to get the configure finishing without errors. The following gmake (GNU make on FreeBSD) resulted in and endlessly running bash process (maybe waiting on a named pipe to open) preceded by "find: lua_modules: No such file or directory" (3 times). FreeBSD 12.0-RELEASE.

P.S. Just running "make" as Simon wrote resulted in a long slew of errors, so FreeBSD make seems incompatible:

line 998: warning: duplicate script for target "%.pdf" ignored
make: "/sile/sile/Makefile" line 995: warning: using previous script for "%.pdf" defined here
make: "/sile/sile/Makefile" line 1050: warning: duplicate script for target "tests/%.expected" ignored
make: "/sile/sile/Makefile" line 1047: warning: using previous script for "tests/%.expected" defined here
make: "/sile/sile/Makefile" line 1062: warning: duplicate script for target "tests/%.actual" ignored
make: "/sile/sile/Makefile" line 1059: warning: using previous script for "tests/%.actual" defined here
make: "/sile/sile/Makefile" line 1073: Could not find
make: Fatal errors encountered -- cannot continue
make: stopped in /sile/sile

**Comment by alerque** [comment]
Thanks for the feedback @bsdzilla. I'm pretty confident at this point we're relying on several GNU tool specific functions. That's probably worth documenting to save people time when compiling and not realizing they need `gmake`.

I'm a little confused as to whether the current state of affairs is acceptable thought. The 2016 era instructions should be mostly unnecessary. You shouldn't need any of those Lua rocks installed at the system level any more (but you do need luarocks itself), and I'm pretty sure you shouldn't need the custom arguments to `./configure`.

Are you is a position to confirm the minimum user steps required for BSD installation?

**Comment by bsdzilla** [comment]
Can't be sure about the luarocks stuff, but the custom arguments to configure are still  definitely needed! With no args it ends with this: 

checking if Lua header version matches 5.2... yes
checking if LUA_VERSION is defined... yes
checking for library containing exp... -lm
checking for library containing dlopen... none required
checking for library containing lua_load... no
configure: error: cannot find Lua libs

I can check later what steps are absolutely needed on FreeBSD 12, using a clean install. But of course there's the issue of not being able to finish the gmake. All I can do is Ctrl-Break it. It always shows this after interrupting, so that's where it seems to stall: 
gmake
find: lua_modules: No such file or directory
find: lua_modules: No such file or directory
find: lua_modules: No such file or directory
^C
bash: cannot open named pipe /tmp//sh-np.b8yCaU for writing: Interrupted system call

**Comment by alerque** [comment]
@bsdzilla I wonder if we are using GNU specific `find` argument syntax. Does BSD's `find` accept multiple locations? The trouble might be this line:

https://github.com/simoncozens/sile/blob/fea257788be42ad4f33f13862b05a555e13bc8ef/Makefile.am#L10

I wonder, dose any incantation of `find dir1 dir2 -type f -print` works at all on BSD?

**Comment by bsdzilla** [comment]
find dir1 dir2 -type f -print works, accepts multiple locations.

**Comment by alerque** [comment]
We seem to have a catch 22 here. The `lua_modules` directory does not exist until you run `make prep`, `make installrocks`, or something else that depends on it. GNU's find doesn't seem to care and just returns what it can find. The value isn't finalized until the second run anyway, at which time the directory will exist. On BSD it seems that it's just hanging?

Or are we actually hung up at a different place and those messages are just the last thing that _worked_?

**Comment by bsdzilla** [comment]
That's it. It's not the find it hangs on (I would have seen it in the process list), it's something beyond that point. But the only running visible in the process list while it hangs is bash. If I knew which bash I could do a set -x to see what it's doing.

**Comment by alerque** [comment]
@bsdzilla Try adding `.SHELLFLAGS = +x` to `Makefile.am` round about line 18.

**Comment by alerque** [comment]
@bsdzilla I would really like to track this down in the near-ish future. I'm pushing to cut another release with all the changes to the build and install system so that package managers can start getting the installation process cleaned up and hopefully we can get back to working on the typesetter and make more frequent releases easier.

Note that one major change in master these days is that you can chose whether to use system installed Lua dependencies or bundle the needed Lua Rocks next to the SILE code:

```sh
# Default
$ ./configure --without-system-luarocks

# Optional
$ ./configure --with-system-luarocks
```

Your issue on BSD with the build process hanging is some other issue and I'd like to get that resolved so that packaging on BSD stands a chance during this release cycle.

**Comment by alerque** [comment]
@udsd007 Are you still around? We've just done a major release with a lot of build related stuff fixed. This _might_ work for you now, and if not I'd be interested in pushing any fixes that are required to make setup easry on BSD through right away in a point release. I'd love any feedback on the status of this.

Also pinging @bsdzilla, same story ;-)

**Comment by bsdzilla** [comment]
still around, will try and let  you know..

**Comment by bsdzilla** [comment]
bootstrap.sh runs ok. After that configure gives:
checking if Lua header version matches 5.2... yes
checking if LUA_VERSION is defined... yes
checking for library containing exp... rm: core: is a directory
rm: core: is a directory
-lm
checking for library containing dlopen... rm: core: is a directory
none required
checking for library containing lua_load... rm: core: is a directory
rm: core: is a directory
rm: core: is a directory
rm: core: is a directory
rm: core: is a directory
rm: core: is a directory
rm: core: is a directory
rm: core: is a directory
rm: core: is a directory
rm: core: is a directory
rm: core: is a directory
no
configure: error: cannot find Lua libs
rm: core: is a directory

**Comment by bsdzilla** [comment]
this is FreeBSD 12.0

**Comment by alerque** [comment]
@bsdzilla I don't know how relevant OpenBSD is here but you might see [this thread](https://marc.info/?t=157907840200001).

**Comment by alerque** [comment]
@bsdzilla The warning about core can be ignored. Not finding Lua stuff is a problem. Is this plain `./configure` or `--with-system-luarocks`?

**Comment by bsdzilla** [comment]
it was without anything done before running, as I assumed it would install all requirements now based on something you said.  Well, that wasn't the case.
Reran Simon's 26 dev 2016 procedure from above. 
Plain configure still gives "configure: error: cannot find Lua libs."
Gmake still hangs somewhere. So basically, nothing has changed...

(adding .SHELLFLAGS = -x to Makefile.am round about line 18 does nothing, causes no extra output. )

**Comment by alerque** [comment]
@bsdzilla Something doesn't smell right here. The default really has changed. Did you start with the v0.10.0 tarball or git? If git I think your working tree is cluttered up with autoconf cacheing. It thinks its a lot smarter that it is. I would nuke with `git clean -dfx` or similar and see if configure comes  to its senses.

**Comment by bsdzilla** [comment]
I used the v0.10.0.tar.gz tarball in a different jail from the last experiment.  I can guarantee that the working tree did not exist before this test.

**Comment by bsdzilla** [comment]
tried ./configure CFLAGS="-I/usr/local/include" LDFLAGS="-L/usr/local/lib" LUA_LIB="-llua-5.2" --with-system-luarocks
It failed ten times or so on missing LUA modules I had to install via luarocks-5.2 install
After that it finished
now gmake does something completely different:
bash: find core classes languages packages lua-libraries -type f -print: No such file or directory
bash: find core classes languages packages lua-libraries -type f -print: No such file or directory
bash: find core classes languages packages lua-libraries -type f -print: No such file or directory
bash: test -d .git && awk '{print ".git/" $2}' .git/HEAD: No such file or directory
bash: git rev-parse --short HEAD: No such file or directory
bash: git rev-parse --short HEAD: No such file or directory
bash: git rev-parse --short ^: No such file or directory
bash: git rev-parse --short HEAD: No such file or directory
bash: git rev-parse --short HEAD: No such file or directory
bash: git rev-parse --short ^: No such file or directory
bash: git rev-parse --short HEAD: No such file or directory
bash: git rev-parse --short HEAD: No such file or directory
bash: git rev-parse --short ^: No such file or directory
bash: git rev-parse --short HEAD: No such file or directory
bash: git rev-parse --short ^: No such file or directory
bash: echo {compare-/,}{tests/bug-465-a.actual,tests/bug-264b.pdf}: No such file or directory
mv .version{,-prev} || touch .version-prev
( test -d .git && ./build-aux/git-version-gen .tarball-version || echo UNKNOWN ) > .version
cmp -s .version{,-prev} || autoreconf configure.ac -f
bash: mv .version{,-prev} || touch .version-prev
( test -d .git && ./build-aux/git-version-gen .tarball-version || echo UNKNOWN ) > .version
cmp -s .version{,-prev} || autoreconf configure.ac -f: No such file or directory
gmake: *** [Makefile:1045: .version] Error 127

**Comment by bsdzilla** [comment]
After seeing all these git errors I cloned the master branch in a new dir, and started with bootstrap and Simon's recipe. This time, it ran all the to the end, and I could do a gmake install! I can start sile and get a prompt.
Turns out the tarball method is unusable...

**Comment by udsd007** [comment]
Hi, Caleb! Thanks for the note. I will give it a full-tilt try, and will keep you & company updated. 

Mike Andrews, W5EGO
WWME Oklahoma area Presenting Team

**Comment by alerque** [comment]
@bsdzilla I really appreciate the testing and feedback here. Not having a BSD box to hack on myself and knowing relatively little about the platform differences is making this a very arduous process. I do still want to get it fixes though.

> Turns out the tarball method is unusable...

This can't be quite true because we're building and installing packages on a number of other platforms from the tarball without a hitch. Something just isn't playing nicely with BSD. I'm still trying to sort out what that is.

I just went through and make some effort to either avoid or suppress anything triggering unnecessary warnings during the configure and compile process. This might make it a lot easier to _see_ what is going on in the output.

Here is a source tarball built with these cleanups (see #792). Normally `make dist` would produce *.tar.bz2, the `make dist-zip` producing a *.zip file is just so that I can attach it here. That should be the only difference from a release tarball:

* [sile-0.10.0.r48-ge022f43.zip](https://github.com/sile-typesetter/sile/files/4108520/sile-0.10.0.r48-ge022f43.zip)

Could you tell me if `./configure --with-system-luarocks && make` runs cleanly? Note of course this requires all the Lua deps installed on your system already.

**Comment by alerque** [comment]
@udsd007 If you had any trouble before you might also see the attachment on my last comment as it will throw fewer distracting warnings about non-issues. If you didn't have trouble and got it to work we'd love hear about it too!

**Comment by alerque** [comment]
Note the above mentioned compile time cleanups are in the v0.10.1 release available from the releases page now. I'd still like to hear the status of building this on BSDs and anything we can do to facilitate that or get the official port up to date.

**Comment by bsdzilla** [comment]
unzipped your sile zip in a dir on the system with the working git-based compile.
./configure --with-system-luarocks && gmake 
result: 
checking for library containing dlopen... none required
checking for library containing lua_load... no
configure: error: cannot find Lua libs

So: same as before.

with ./configure CFLAGS="-I/usr/local/include" LDFLAGS="-L/usr/local/lib" LUA_LIB="-llua-5.2"
it finishes configure ok, but hangs on the gmake, after saying: 
checking for library containing dlopen... none required
checking for library containing lua_load... no
configure: error: cannot find Lua libs

So this zip file doesn't give a substantial difference from the tarball, while a git clone works just fine.

**Comment by simoncozens** [comment]
Just dropping this here for now: https://cirrus-ci.org/guide/FreeBSD/
Will try to set it up later; kids off sick today and lots of deadlines for dayjob stuff.

**Comment by alerque** [comment]
@bsdzilla Can you try running `autoreconf --force` before `./configure ...`? This is starting to look like something the bootstrap script is getting done in adjusting for platforms is not getting done in the dist build. If it's not that, maybe the `glibtoolize` step or one of the others is needed.

When we figure out what that is, I'm guessing we'll see what's wrong with the way we're using automake stuff...

**Comment by alerque** [comment]
I think the bulk of the remaining issues here should be fixed as of #804 which fixed various make issues that Debian was having. I'm leaving this open until we get confirmation form BSD folks that the tarball install goes better!

**Comment by alerque** [comment]
We're regularly testing CI builds on FreeBSD (12) now, any specific issues end users have should now be opened separately. I think this tracking issue has jumped the shark.
